src/libraries/System.Threading.RateLimiting/src/System/Threading/RateLimiting/PartitionedRateLimiter.cs

```diff
@@ -182,48 +184,77 @@ private void ThrowIfDisposed()
-                    foreach (KeyValuePair<TKey, Lazy<RateLimiter>> kvp in limiter._limiters)
+                    _cachedLimiters.Clear();
+                    _cachedLimiters = _limiters.ToList();
+                }
+            }
+
+            List<RateLimiter>? limitersToDispose = null;
+
+            // cachedLimiters is safe to use outside the lock because it is only updated by the Timer
+            foreach (KeyValuePair<TKey, Lazy<RateLimiter>> rateLimiter in _cachedLimiters)
+            {
+                if (!rateLimiter.Value.IsValueCreated)
+                {
+                    continue;
+                }
+                if (rateLimiter.Value.Value.IdleDuration is TimeSpan idleDuration && idleDuration > _idleTimeLimit)
+                {
+                    lock (Lock)
                     {
-                        if (kvp.Value.IsValueCreated)
-                        {
-                            if (kvp.Value.Value is ReplenishingRateLimiter)
-                            {
-                                limiter._cachedLimiters.Add(kvp.Value);
-                            }
-                        }
-                        else
+                        // Check time again under lock to make sure no one calls Acquire or WaitAsync after checking the time and removing the limiter
+                        idleDuration = rateLimiter.Value.Value.IdleDuration ?? TimeSpan.Zero;
+                        if (idleDuration > _idleTimeLimit)
                         {
-                            // In rare cases the RateLimiter will be added to the storage but not be initialized yet
-                            // keep cache invalid if there was a non-initialized RateLimiter
-                            // the next time we run the timer the cache will be updated
-                            // with the initialized RateLimiter
-                            cacheStillInvalid = true;
+                            // Remove limiter from the lookup table and mark cache as invalid
+                            // If a request for this partition comes in it will have to create a new limiter now
+                            // And the next time the timer runs the cache needs to be updated to no longer have a reference to this limiter
+                            _cacheInvalid = true;
+                            _limiters.Remove(rateLimiter.Key);
+
+                            // We don't want to dispose inside the lock so we need to defer it
+                            limitersToDispose ??= new List<RateLimiter>();
+                            limitersToDispose.Add(rateLimiter.Value.Value);
                         }
                     }
-                    limiter._cacheInvalid = cacheStillInvalid;
+                }
+                else if (rateLimiter.Value.Value is ReplenishingRateLimiter replenishingRateLimiter)
+                {
+                    replenishingRateLimiter.TryReplenish();
```

Comment:
It looks like if one of the non-expired limiters throws from `TryReplenish`, we never dispose the limiters that expired that iteration of the `TimerLoop()`.